### PR TITLE
Implement constrained click and drag that doesn't leave spaces

### LINKED ISSUE #240: [FEAT] Left-click and drag creation shouldn't leave gaps, and should constraint to horizontal/vertical
I often find myself drawing belts like this:

image

If you left click and move the mouse too quickly, it skips spaces. Also, if you are trying to move horizontal, it's easy to 'drift' vertical (as shown above) which is rarely what you want.

Factorio handles this well - dragging belts or other items constraints to horizontal and fills in every square. I'd like to replicate that behavior - constrain painting to horizontal or vertical and ensure every square is filled in.

This is especially powerful when zoomed out - before it was very hard to draw a continuous belt, now it's easy. And FBE is better than factorio in this as there is no need for vision to work on large blueprints.

Note that right click and drag will also skip items. I thought about implementing the feature for it, but you don't want right click to be horizontal/vertically constrained. So I didn't implement it for that; I think it's fine because you have ctrl+right click do delete a region.

I have a working PR that implements this.

### PR DESCRIPTION
**Pull request by KenCoder**
Closes #240

### CODE REVIEW

**Review by teoxoy — CHANGES_REQUESTED** [review]
Thanks for the PR!

Can we instead move the constrain logic into the `draggingCreateUpdateFn` listener and in GridData only fire the inbetween `update16` and `update32` events (something like the code below)? For this to work properly I think we need to make use of the x and y args in the listeners in some places.

```ts
// emit update32 when mouse changes tile whithin the 32 pixel size grid
if (!(oldX32 === this._x32 && oldY32 === this._y32)) {
    const xDelta = this._x32 - oldX32
    const xAbs = Math.abs(xDelta)
    const xSign = Math.sign(xDelta)
    const yDelta = this._y32 - oldY32
    const yAbs = Math.abs(yDelta)
    const ySign = Math.sign(yDelta)
    for (let i = 0; i < xAbs; i++) {
        for (let j = 0; j < yAbs; j++) {
            this.emit('update32', this._x32 + i * xSign, this._y32 + j * ySign)
        }
    }
}
```

One last thing: There is some weird behavior now with the mouse hover. After clicking anywhere, you can't hover over entities anymore. Not sure why this is happening.

**Comment by KenCoder** [comment]
I first had the constrain logic inside draggingCreateUpdateFn , but the problem is that the paint container takes its location directly from gridData, so all the position related things (like the hovering cursor) are wrong. It would also require modifying placeEntityContainer everywhere to take an optional position. And even then, things like checkBuildable use the current position so I worry it would require a lot of tweaking.  That seemed uglier and riskier but I can bring it back if you want to see what it's like. 

Also for the update logic it seemed like it might matter that the x16, y16, x32 and y32 variables be consistent.  Your code does not step the x16 variables.  Note that even though the positions are passed as part of the update event, many handlers look back at gridData to get the position in the handler (this code for example):

```
    protected setNewPosition(size: IPoint): void {
        if (size.x % 2 === 0) {
            const npx = this.bpc.gridData.x16 * 16
            this.x = npx + (npx % 32 === 0 ? 0 : 16)
        } else {
            this.x = this.bpc.gridData.x32 * 32 + 16
        }

        if (size.y % 2 === 0) {
            const npy = this.bpc.gridData.y16 * 16
            this.y = npy + (npy % 32 === 0 ? 0 : 16)
        } else {
            this.y = this.bpc.gridData.y32 * 32 + 16
        }
    }
```

My idea was that if it "looked" to the all code like the user just happened to drag the mouse in a horizontal line that is slow enough to hit every point, then everything would just work (which it did). Having gridData made this super easy.

File: packages/editor/src/containers/BlueprintContainer.ts [hunk 1]
```diff
@@ -35,6 +35,8 @@ export enum EditorMode {
     COPY,
     /** Active when selecting multiple entities for deletion */
     DELETE,
+    /** Active while LButton or Rbutton down, to place / delete entities in constrained line */
+    ACTING,
```
[the remark this answers is not in the document]
Comment (KenCoder) [reply packages/editor/src/containers/BlueprintContainer.ts]: Oops - fixed thanks.

**Comment by KenCoder** [comment]
I fixed the issue with the editor not opening - thanks!

File: packages/editor/src/containers/GridData.ts [hunk 2]
```diff
@@ -70,27 +98,45 @@ export class GridData extends EventEmitter {
...
+        while (more) {
+            more = this.constrainMove('_x16',  x16) || this.constrainMove('_y16',  y16) ||
+                   this.constrainMove('_x32',  x32) || this.constrainMove('_y32',  y32)
+
+            if (this.bpc.mode === EditorMode.PAN) return
+
+            // emit update when mouse changes tile whithin the 1 pixel size grid
+            if (!(oldX === this._x && oldY === this._y)) {
+                this.emit('update', this._x, this._y)
+            }
```
Comment (teoxoy) [inline packages/editor/src/containers/GridData.ts:131]: Can we move these lines out of the loop?

File: packages/editor/src/containers/GridData.ts [hunk 3]
```diff
@@ -70,27 +98,45 @@ export class GridData extends EventEmitter {
...
         }
-        // emit update32 when mouse changes tile whithin the 32 pixel size grid
-        if (!(oldX32 === this._x32 && oldY32 === this._y32)) {
-            this.emit('update32', this._x32, this._y32)
+
+        let more = true
+        while (more) {
+            more = this.constrainMove('_x16',  x16) || this.constrainMove('_y16',  y16) ||
+                   this.constrainMove('_x32',  x32) || this.constrainMove('_y32',  y32)
```
Comment (teoxoy) [inline packages/editor/src/containers/GridData.ts:124]: Can we have just one function call here and do the checking internally?

**Review by teoxoy — APPROVED** [review]
Thanks for the insight! I agree it might be too involved to change how we read those coordinates (at least for now).

Besides the 2 more comments I left, seems to be working properly now.
I've been playing around with the feature and I quite like it already, good stuff!

File: packages/editor/src/containers/GridData.ts (hunk 2 above)
Comment (KenCoder) [reply packages/editor/src/containers/GridData.ts:131]: Yes good idea

File: packages/editor/src/containers/GridData.ts (hunk 3 above)
Comment (KenCoder) [reply packages/editor/src/containers/GridData.ts]: I updated to use an intermediate function, also noticed prior logic was slightly off.

File: packages/editor/src/containers/GridData.ts (hunk 2 above)
Comment (teoxoy) [reply packages/editor/src/containers/GridData.ts]: The pan mode check should also go out (before emitting any events).

Comment (KenCoder) [reply packages/editor/src/containers/GridData.ts]: It needs to go after the update, though. I could repeat the update inside the pan check if you like that better.

Comment (teoxoy) [reply packages/editor/src/containers/GridData.ts]: It should be fine actually since while panning all the values will (should) remain the same.

Comment (teoxoy) [reply packages/editor/src/containers/GridData.ts]: The check can even be moved all the way up where we check if we have a `this.bpc`

Comment (KenCoder) [reply packages/editor/src/containers/GridData.ts]: Ok, done

**Review by teoxoy — APPROVED** [review]

**Pull request merged**